classes/engine.php

```diff
@@ -0,0 +1,456 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Elasticsearch engine.
+ *
+ * Provides an interface between Moodles Global search functionality
+ * and the Elasticsearch (https://www.elastic.co/products/elasticsearch)
+ * search engine.
+ *
+ * Elasticsearch presents a REST Webservice API that we communicate with
+ * via Curl.
+ *
+ * @package     search_elastic
+ * @copyright   Matt Porritt <mattp@catalyst-au.net>
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace search_elastic;
+
+defined('MOODLE_INTERNAL') || die();
+
+class engine extends \core_search\engine {
```

Comment:
Will see what I can do here, might be hard ot make class simpler based on requirements of API